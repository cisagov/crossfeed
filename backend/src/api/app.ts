import * as express from 'express';
import * as bodyParser from 'body-parser';
import * as cookieParser from 'cookie-parser';
import * as cookie from 'cookie';
import * as cors from 'cors';
import * as helmet from 'helmet';
import { handler as healthcheck } from './healthcheck';
import { handler as scheduler } from '../tasks/scheduler';
import * as auth from './auth';
import * as domains from './domains';
import * as search from './search';
import * as vulnerabilities from './vulnerabilities';
import * as organizations from './organizations';
import * as scans from './scans';
import * as users from './users';
import * as scanTasks from './scan-tasks';
import * as stats from './stats';
import { listenForDockerEvents } from './docker-events';
import { createProxyMiddleware } from 'http-proxy-middleware';

if (
  (process.env.IS_OFFLINE || process.env.IS_LOCAL) &&
  typeof jest === 'undefined'
) {
  listenForDockerEvents();

  setInterval(() => scheduler({}, {} as any, () => null), 30000);
}

const handlerToExpress = (handler) => async (req, res, next) => {
  const { statusCode, body } = await handler(
    {
      pathParameters: req.params,
      requestContext: req.requestContext,
      body: JSON.stringify(req.body || '{}'),
      headers: req.headers,
      path: req.originalUrl
    },
    {}
  );
  try {
    const parsedBody = JSON.parse(body);
    res.status(statusCode).json(parsedBody);
  } catch (e) {
    // Not a JSON body
    res.setHeader('content-type', 'text/plain');
    res.status(statusCode).send(body);
  }
};

const app = express();

app.use(cors());
app.use(bodyParser.json());
app.use(helmet.hsts());
app.use(cookieParser());

app.get('/', handlerToExpress(healthcheck));
app.post('/auth/login', handlerToExpress(auth.login));
app.post('/auth/callback', handlerToExpress(auth.callback));

const checkUserLoggedIn = async (req, res, next) => {
  req.requestContext = {
    authorizer: await auth.authorize({
      authorizationToken: req.headers.authorization
    })
  };
  if (
    !req.requestContext.authorizer.id ||
    req.requestContext.authorizer.id === 'cisa:crossfeed:anonymous'
  ) {
    return res.status(401).send('Not logged in');
  }
  return next();
};

const checkUserSignedTerms = (req, res, next) => {
  // Bypass ToU for CISA emails
  const approvedEmailAddresses = ['@cisa.dhs.gov'];
  if (process.env.NODE_ENV === 'test')
    approvedEmailAddresses.push('@crossfeed.cisa.gov');
  for (const email of approvedEmailAddresses) {
    if (
      req.requestContext.authorizer.email &&
      req.requestContext.authorizer.email.endsWith(email)
    )
      return next();
  }
  if (
    !req.requestContext.authorizer.dateAcceptedTerms ||
    (req.requestContext.authorizer.acceptedTermsVersion &&
      req.requestContext.authorizer.acceptedTermsVersion !==
        getToUVersion(req.requestContext.authorizer))
  ) {
    return res.status(403).send('User must accept terms of use');
  }
  return next();
};

const getMaximumRole = (user) => {
  if (user?.userType === 'globalView') return 'user';
  return user && user.roles && user.roles.find((role) => role.role === 'admin')
    ? 'admin'
    : 'user';
};

const getToUVersion = (user) => {
  return `v${process.env.REACT_APP_TERMS_VERSION}-${getMaximumRole(user)}`;
};

// Rewrite the URL for some Matomo admin dashboard URLs, due to a bug in
// how Matomo handles relative URLs when hosted on a subpath.
app.get('/plugins/Morpheus/images/logo.svg', (req, res) =>
  res.redirect('/matomo/plugins/Morpheus/images/logo.svg?matomo')
);
app.get('/index.php', (req, res) => res.redirect('/matomo/index.php'));

const matomoProxy = createProxyMiddleware({
  target: process.env.MATOMO_URL,
  headers: { HTTP_X_FORWARDED_URI: '/matomo' },
  pathRewrite: function (path, req) {
    return path.replace(/^\/matomo/, '');
  },
  onProxyReq: function (proxyReq, req, res) {
    // Only pass the MATOMO_SESSID cookie to Matomo.
    const cookies = cookie.parse(proxyReq.getHeader('Cookie'));
    const newCookies = cookie.serialize(
      'MATOMO_SESSID',
      String(cookies['MATOMO_SESSID'])
    );
    proxyReq.setHeader('Cookie', newCookies);
  },
  onProxyRes: function (proxyRes, req, res) {
    // Remove transfer-encoding: chunked responses, because API Gateway doesn't
    // support chunked encoding.
    if (proxyRes.headers['transfer-encoding'] === 'chunked') {
      proxyRes.headers['transfer-encoding'] = '';
    }
  }
});

app.use(
  '/matomo',
  async (req, res, next) => {
    // Public paths -- see https://matomo.org/docs/security-how-to/
    const ALLOWED_PATHS = ['/matomo.php', '/matomo.js'];
    if (ALLOWED_PATHS.indexOf(req.path) > -1) {
      return next();
    }
    // API Gateway isn't able to proxy fonts properly -- so we're using a CDN instead.
    if (req.path === '/plugins/Morpheus/fonts/matomo.woff2') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.woff2'
      );
    }
    if (req.path === '/plugins/Morpheus/fonts/matomo.woff') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.woff'
      );
    }
    if (req.path === '/plugins/Morpheus/fonts/matomo.ttf') {
      return res.redirect(
        'https://cdn.jsdelivr.net/gh/matomo-org/matomo@3.14.1/plugins/Morpheus/fonts/matomo.ttf'
      );
    }
    // Only allow global admins to access all other paths.
    const user = (await auth.authorize({
      authorizationToken: req.cookies['crossfeed-token']
    })) as auth.UserToken;
    if (user.userType !== 'globalAdmin') {
      return res.status(401).send('Unauthorized');
    }
    return next();
  },
  matomoProxy
);

// Routes that require an authenticated user, without
// needing to sign the terms of service yet
const authenticatedNoTermsRoute = express.Router();
authenticatedNoTermsRoute.use(checkUserLoggedIn);
authenticatedNoTermsRoute.get('/users/me', handlerToExpress(users.me));
authenticatedNoTermsRoute.post(
  '/users/me/acceptTerms',
  handlerToExpress(users.acceptTerms)
);
authenticatedNoTermsRoute.put('/users/:userId', handlerToExpress(users.update));

app.use(authenticatedNoTermsRoute);

// Routes that require an authenticated user that has
// signed the terms of service
const authenticatedRoute = express.Router();
authenticatedRoute.use(checkUserLoggedIn);
authenticatedRoute.use(checkUserSignedTerms);

authenticatedRoute.post('/search', handlerToExpress(search.search));
authenticatedRoute.post('/domain/search', handlerToExpress(domains.list));
authenticatedRoute.post('/domain/export', handlerToExpress(domains.export_));
authenticatedRoute.get('/domain/:domainId', handlerToExpress(domains.get));
authenticatedRoute.post(
  '/vulnerabilities/search',
  handlerToExpress(vulnerabilities.list)
);
authenticatedRoute.post(
  '/vulnerabilities/export',
  handlerToExpress(vulnerabilities.export_)
);
authenticatedRoute.get(
  '/vulnerabilities/:vulnerabilityId',
  handlerToExpress(vulnerabilities.get)
);
authenticatedRoute.put(
  '/vulnerabilities/:vulnerabilityId',
  handlerToExpress(vulnerabilities.update)
);
authenticatedRoute.get('/scans', handlerToExpress(scans.list));
authenticatedRoute.get('/granularScans', handlerToExpress(scans.listGranular));
authenticatedRoute.post('/scans', handlerToExpress(scans.create));
authenticatedRoute.get('/scans/:scanId', handlerToExpress(scans.get));
authenticatedRoute.put('/scans/:scanId', handlerToExpress(scans.update));
authenticatedRoute.delete('/scans/:scanId', handlerToExpress(scans.del));
authenticatedRoute.post('/scans/:scanId/run', handlerToExpress(scans.runScan));
authenticatedRoute.post(
  '/scheduler/invoke',
  handlerToExpress(scans.invokeScheduler)
);
authenticatedRoute.post('/scan-tasks/search', handlerToExpress(scanTasks.list));
authenticatedRoute.post(
  '/scan-tasks/:scanTaskId/kill',
  handlerToExpress(scanTasks.kill)
);
authenticatedRoute.get(
  '/scan-tasks/:scanTaskId/logs',
  handlerToExpress(scanTasks.logs)
);

authenticatedRoute.get('/organizations', handlerToExpress(organizations.list));
authenticatedRoute.get(
  '/organizations/public',
  handlerToExpress(organizations.listPublicNames)
);
authenticatedRoute.get(
  '/organizations/:organizationId',
  handlerToExpress(organizations.get)
);
authenticatedRoute.post(
  '/organizations',
  handlerToExpress(organizations.create)
);
authenticatedRoute.put(
  '/organizations/:organizationId',
  handlerToExpress(organizations.update)
);
authenticatedRoute.delete(
  '/organizations/:organizationId',
  handlerToExpress(organizations.del)
);
authenticatedRoute.post(
  '/organizations/:organizationId/roles/:roleId/approve',
  handlerToExpress(organizations.approveRole)
);
authenticatedRoute.post(
  '/organizations/:organizationId/roles/:roleId/remove',
  handlerToExpress(organizations.removeRole)
);
authenticatedRoute.post(
  '/organizations/:organizationId/granularScans/:scanId/update',
  handlerToExpress(organizations.updateScan)
);
authenticatedRoute.post('/stats', handlerToExpress(stats.get));
authenticatedRoute.get('/users', handlerToExpress(users.list));
authenticatedRoute.post('/users', handlerToExpress(users.invite));
authenticatedRoute.delete('/users/:userId', handlerToExpress(users.del));

app.use(authenticatedRoute);

export default app;
