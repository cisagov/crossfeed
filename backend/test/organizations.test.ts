import * as request from 'supertest';
import app from '../src/api/app';
import { createUserToken, DUMMY_USER_ID } from './util';
import {
  Organization,
  Role,
  connectToDatabase,
  Scan,
  ScanTask,
  User
} from '../src/models';

describe('organizations', () => {
  beforeAll(async () => {
    await connectToDatabase();
  });
  describe('create', () => {
    it('create by globalAdmin should succeed', async () => {
      const user = await User.create({
        firstName: '',
        lastName: '',
        email: Math.random() + '@crossfeed.cisa.gov',
        userType: 'globalAdmin'
      }).save();
      const name = 'test-' + Math.random();
      const response = await request(app)
        .post('/organizations/')
        .set(
          'Authorization',
          createUserToken({
            id: user.id,
            userType: 'globalAdmin'
          })
        )
        .send({
          ipBlocks: [],
          name,
          rootDomains: ['cisa.gov'],
          isPassive: false,
          inviteOnly: true
        })
        .expect(200);
      expect(response.body).toMatchSnapshot({
        createdAt: expect.any(String),
        updatedAt: expect.any(String),
        id: expect.any(String),
        name: expect.any(String),
        createdBy: {
          id: expect.any(String)
        }
      });
      expect(response.body.createdBy.id).toEqual(user.id);
      expect(response.body.name).toEqual(name);
    });
    it("can't add organization with the same name", async () => {
      const user = await User.create({
        firstName: '',
        lastName: '',
        email: Math.random() + '@crossfeed.cisa.gov',
        userType: 'globalAdmin'
      }).save();
      const name = 'test-' + Math.random();
      await request(app)
        .post('/organizations/')
        .set(
          'Authorization',
          createUserToken({
            id: user.id,
            userType: 'globalAdmin'
          })
        )
        .send({
          ipBlocks: [],
          name,
          rootDomains: ['cisa.gov'],
          isPassive: false,
          inviteOnly: true
        });
      // .expect(200);
      const response = await request(app)
        .post('/organizations/')
        .set(
          'Authorization',
          createUserToken({
            id: user.id,
            userType: 'globalAdmin'
          })
        )
        .send({
          ipBlocks: [],
          name,
          rootDomains: ['cisa.gov'],
          isPassive: false,
          inviteOnly: true
        })
        .expect(500);
      expect(response.body).toMatchSnapshot();
    });
    it('create by globalView should fail', async () => {
      const name = 'test-' + Math.random();
      const response = await request(app)
        .post('/organizations/')
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalView'
          })
        )
        .send({
          ipBlocks: [],
          name,
          rootDomains: ['cisa.gov'],
          isPassive: false,
          inviteOnly: true
        })
        .expect(403);
      expect(response.body).toEqual({});
    });
  });
  describe('update', () => {
    it('update by globalAdmin should succeed', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const name = 'test-' + Math.random();
      const rootDomains = ['test-' + Math.random()];
      const ipBlocks = ['1.1.1.1'];
      const isPassive = true;
      const inviteOnly = false;
      const response = await request(app)
        .put(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalAdmin'
          })
        )
        .send({
          name,
          rootDomains,
          ipBlocks,
          isPassive,
          inviteOnly
        })
        .expect(200);
      expect(response.body.name).toEqual(name);
      expect(response.body.rootDomains).toEqual(rootDomains);
      expect(response.body.ipBlocks).toEqual(ipBlocks);
      expect(response.body.isPassive).toEqual(isPassive);
      expect(response.body.inviteOnly).toEqual(inviteOnly);
    });
    it('update by org admin should update everything but rootDomains and ipBlocks', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const name = 'test-' + Math.random();
      const rootDomains = ['test-' + Math.random()];
      const ipBlocks = ['1.1.1.1'];
      const isPassive = true;
      const inviteOnly = false;
      const response = await request(app)
        .put(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .send({
          name,
          rootDomains,
          ipBlocks,
          isPassive,
          inviteOnly
        })
        .expect(200);
      expect(response.body.name).toEqual(name);
      expect(response.body.rootDomains).toEqual(organization.rootDomains);
      expect(response.body.ipBlocks).toEqual(organization.ipBlocks);
      expect(response.body.isPassive).toEqual(isPassive);
      expect(response.body.inviteOnly).toEqual(inviteOnly);
    });
    it('update by globalView should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const name = 'test-' + Math.random();
      const rootDomains = ['test-' + Math.random()];
      const ipBlocks = ['1.1.1.1'];
      const isPassive = true;
      const inviteOnly = false;
      const response = await request(app)
        .put(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalView'
          })
        )
        .send({
          name,
          rootDomains,
          ipBlocks,
          isPassive,
          inviteOnly
        })
        .expect(403);
      expect(response.body).toEqual({});
    });
  });
  describe('delete', () => {
    it('delete by globalAdmin should succeed', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .delete(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalAdmin'
          })
        )
        .expect(200);
      expect(response.body.affected).toEqual(1);
    });
    it('delete by org admin should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .delete(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
    it('delete by globalView should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .delete(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalView'
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
  });
  describe('list', () => {
    it('list by globalView should succeed', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .get(`/organizations`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalView'
          })
        )
        .expect(200);
      expect(response.body.length).toBeGreaterThanOrEqual(1);
    });
    it('list by org member should only get their org', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      // this org should not show up in the response
      await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .get(`/organizations`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'user'
              }
            ]
          })
        )
        .expect(200);
      expect(response.body.length).toEqual(1);
      expect(response.body[0].id).toEqual(organization.id);
    });
  });
  describe('listPublicNames', () => {
    it('listPublicNames by non-org member should succeed', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false,
        inviteOnly: false
      }).save();
      const response = await request(app)
        .get(`/organizations/public`)
        .set('Authorization', createUserToken({}))
        .expect(200);
      expect(response.body.length).toBeGreaterThanOrEqual(1);
      expect(
        response.body.map((e) => e.id).indexOf(organization.id)
      ).not.toEqual(-1);
    });
    it('listPublicNames with bad auth key should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false,
        inviteOnly: false
      }).save();
      const response = await request(app)
        .get(`/organizations/public`)
        .set('Authorization', '')
        .expect(401);
      expect(response.body).toEqual({});
    });
  });
  describe('get', () => {
    it('get by globalView should fail - TODO should we change this?', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .get(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalView'
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
    it('get by an org admin user should pass', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .get(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .expect(200);
      expect(response.body.name).toEqual(organization.name);
    });
    it('get by an org admin user of a different org should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const organization2 = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .get(`/organizations/${organization2.id}`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
    it('get by an org regular user should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const response = await request(app)
        .get(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'user'
              }
            ]
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
    it('get by an org admin user should return associated scantasks', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const scan = await Scan.create({
        name: 'censys',
        arguments: {},
        frequency: 999999
      }).save();
      const scanTask = await ScanTask.create({
        scan,
        status: 'created',
        type: 'fargate',
        organization
      }).save();
      const response = await request(app)
        .get(`/organizations/${organization.id}`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .expect(200);
      expect(response.body.name).toEqual(organization.name);
      expect(response.body.scanTasks.length).toEqual(1);
      expect(response.body.scanTasks[0].id).toEqual(scanTask.id);
      expect(response.body.scanTasks[0].scan.id).toEqual(scan.id);
    });
  });
  describe('update', () => {
    it('enabling a scan by org admin for an organization should succeed', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const scan = await Scan.create({
        name: 'censys',
        arguments: {},
        frequency: 999999,
        isGranular: true
      }).save();
      const response = await request(app)
        .post(
          `/organizations/${organization.id}/granularScans/${scan.id}/update`
        )
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .send({
          enabled: true
        })
        .expect(200);
      expect(response.body.granularScans.length).toEqual(1);
      const updated = (await Organization.findOne(
        {
          id: organization.id
        },
        {
          relations: ['granularScans']
        }
      )) as Organization;
      expect(updated.name).toEqual(organization.name);
      expect(updated.granularScans).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            id: scan.id
          })
        ])
      );
    });
    it('disabling a scan by org admin for an organization should succeed', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const scan = await Scan.create({
        name: 'censys',
        arguments: {},
        frequency: 999999,
        organizations: [organization],
        isGranular: true
      }).save();
      const response = await request(app)
        .post(
          `/organizations/${organization.id}/granularScans/${scan.id}/update`
        )
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .send({
          enabled: false
        })
        .expect(200);
      const updated = (await Organization.findOne(
        {
          id: organization.id
        },
        {
          relations: ['granularScans']
        }
      )) as Organization;
      expect(updated.name).toEqual(organization.name);
      expect(updated.granularScans).toEqual([]);
    });
    it('enabling a scan by org user for an organization should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const scan = await Scan.create({
        name: 'censys',
        arguments: {},
        frequency: 999999
      }).save();
      const response = await request(app)
        .post(
          `/organizations/${organization.id}/granularScans/${scan.id}/update`
        )
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'user'
              }
            ]
          })
        )
        .send({
          enabled: true
        })
        .expect(403);
    });
    it('enabling a scan by globalAdmin for an organization should succeed', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const scan = await Scan.create({
        name: 'censys',
        arguments: {},
        frequency: 999999,
        isGranular: true
      }).save();
      const response = await request(app)
        .post(
          `/organizations/${organization.id}/granularScans/${scan.id}/update`
        )
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalAdmin'
          })
        )
        .send({
          enabled: true
        })
        .expect(200);
      const updated = (await Organization.findOne(
        {
          id: organization.id
        },
        {
          relations: ['granularScans']
        }
      )) as Organization;
      expect(updated.name).toEqual(organization.name);
      expect(updated.granularScans).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            id: scan.id
          })
        ])
      );
    });
  });
  describe('approveRole', () => {
    it('approveRole by globalAdmin should work', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      let role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/approve`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalAdmin'
          })
        )
        .expect(200);
      expect(response.body).toEqual({});

      role = (await Role.findOne(role.id, {
        relations: ['approvedBy']
      })) as Role;
      expect(role.approved).toEqual(true);
      expect(role.approvedBy.id).toEqual(DUMMY_USER_ID);
    });
    it('approveRole by globalView should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/approve`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalView'
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
    it('approveRole by org admin should work', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/approve`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .expect(200);
      expect(response.body).toEqual({});
    });
    it('approveRole by org user should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/approve`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'user'
              }
            ]
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
  });
  describe('removeRole', () => {
    it('removeRole by globalAdmin should work', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/remove`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalAdmin'
          })
        )
        .expect(200);
      expect(response.body.affected).toEqual(1);
    });
    it('removeRole by globalView should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/remove`)
        .set(
          'Authorization',
          createUserToken({
            userType: 'globalView'
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
    it('removeRole by org admin should work', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/remove`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'admin'
              }
            ]
          })
        )
        .expect(200);
      expect(response.body.affected).toEqual(1);
    });
    it('removeRole by org user should fail', async () => {
      const organization = await Organization.create({
        name: 'test-' + Math.random(),
        rootDomains: ['test-' + Math.random()],
        ipBlocks: [],
        isPassive: false
      }).save();
      const role = await Role.create({
        role: 'user',
        approved: false,
        organization
      }).save();
      const response = await request(app)
        .post(`/organizations/${organization.id}/roles/${role.id}/remove`)
        .set(
          'Authorization',
          createUserToken({
            roles: [
              {
                org: organization.id,
                role: 'user'
              }
            ]
          })
        )
        .expect(403);
      expect(response.body).toEqual({});
    });
  });
});
