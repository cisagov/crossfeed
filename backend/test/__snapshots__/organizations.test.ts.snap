// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organizations create can't add organization with the same name 1`] = `Object {}`;

exports[`organizations create create by globalAdmin should succeed 1`] = `
Object {
  "createdAt": Any<String>,
  "createdBy": Object {
    "id": Any<String>,
  },
  "id": Any<String>,
  "inviteOnly": true,
  "ipBlocks": Array [],
  "isPassive": false,
  "name": Any<String>,
  "rootDomains": Array [
    "cisa.gov",
  ],
  "updatedAt": Any<String>,
}
`;
